Release notes for Pylake v1.0.0 (#480)
lumicks/pylake: changelog.md

# Changelog

## v1.0.0 | 2023-03-14

Happy π-day!

#### New features

> * Added links to example data for all [tutorials](https://lumicks-pylake.readthedocs.io/en/v1.0.0/tutorial/index.html). Try running [them](https://lumicks-pylake.readthedocs.io/en/v1.0.0/tutorial/index.html) in a jupyter notebook!

Reviewer: Maybe just the one link, I found myself trying to figure out what the difference between the two was

```suggestion
* Added links to example data for all [tutorials](https://lumicks-pylake.readthedocs.io/en/v1.0.0/tutorial/index.html). Try running them in a `jupyter` notebook!
```

* Added API to create a [`Kymo`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.kymo.Kymo.html) from an [`ImageStack`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.ImageStack.html). See [`ImageStack.to_kymo()`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.ImageStack.html#lumicks.pylake.ImageStack.to_kymo) for more information.
* Added API for accessing notes in a h5 file, i.e. [`file.notes`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.File.html#lumicks.pylake.File.notes) returns a dictionary of notes.

> * Implemented bias correction for centroid refinement that shrinks the window to reduce estimation bias. This bias correction can be toggled by passing a `bias_correction` argument to [`lk.track_greedy()`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.track_greedy.html#) and [`lk.refine_tracks_centroid()`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.refine_tracks_centroid.html).

Reviewer: maybe add a note that this is default behavior

Author: Good call.

> * Implemented calibration for [`ImageStack`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.ImageStack.html). Images are now plotted with microns on the axes for TIF files where the pixel size is available in the metadata. If the metadata is unavailable, the axes units are provided in pixels.

Reviewer: ```suggestion
* Implemented spatial calibration for [`ImageStack`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.ImageStack.html). Images are now plotted in units of microns for TIF files which contain pixel size in the metadata. If the metadata is unavailable, the axes units are provided in pixels.
```

> * Added option to add a scale bar to by providing a [`lk.ScaleBar()`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.ScaleBar.html) to plotting or export functions. See [plotting and exporting](https://lumicks-pylake.readthedocs.io/en/v1.0.0/tutorial/imagestack.html#plotting-and-exporting) for an example.

Reviewer: ```suggestion
* Added option to show scale bars to image plots using [`lk.ScaleBar()`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.ScaleBar.html). See [plotting and exporting](https://lumicks-pylake.readthedocs.io/en/v1.0.0/tutorial/imagestack.html#plotting-and-exporting) for an example.
```

* Added [`KymoTrackGroup.remove()`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.kymotracker.kymotrack.KymoTrackGroup.html#lumicks.pylake.kymotracker.kymotrack.KymoTrackGroup.remove) to remove a [`KymoTrack`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.kymotracker.kymotrack.KymoTrack.html) from a [`KymoTrackGroup`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.kymotracker.kymotrack.KymoTrackGroup.html).

> * Allow boolean array indexing (e.g. `tracks[[False, False, True]]`) and indexing with arrays of indices (e.g. `tracks[[1, 3]]`) for `KymoTrackGroup`. See the [API documentation](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.kymotracker.kymotrack.KymoTrackGroup.html#lumicks.pylake.kymotracker.kymotrack.KymoTrackGroup) for more information.

Reviewer: ```suggestion
* Enabled boolean array indexing (e.g. `tracks[[False, False, True]]`) and indexing with arrays of indices (e.g. `tracks[[1, 3]]`) for `KymoTrackGroup`. See the [API documentation](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.kymotracker.kymotrack.KymoTrackGroup.html#lumicks.pylake.kymotracker.kymotrack.KymoTrackGroup) for more information.
```

* Improved force calibration documentation. See the new [Tutorial](https://lumicks-pylake.readthedocs.io/en/shuffle_force_calibration/tutorial/force_calibration.html) and [Theory section](https://lumicks-pylake.readthedocs.io/en/shuffle_force_calibration/theory/force_calibration/force_calibration.html).
* Added API to look up (pixel)size of an [`ImageStack`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.ImageStack.html). See [`ImageStack`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.ImageStack.html) for more information.

> * Allow forwarding plot style arguments to [`Slice.range_selector()`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.channel.Slice.html#lumicks.pylake.channel.Slice.range_selector). Prior to this change, a plot with markers was always opened which resulted in very poor performance when plotting high frequency data.

Reviewer: ```suggestion
* Enabled forwarding plot style arguments to [`Slice.range_selector()`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.channel.Slice.html#lumicks.pylake.channel.Slice.range_selector). Prior to this change, a plot with markers was always opened which resulted in very poor performance when plotting high frequency data.
```

#### Breaking changes

* Pylake now requires Python version `>= v3.9` and matplotlib version `>= v3.5`
* When performing particle tracking on kymographs, bias correction is now enabled by default; without this correction, kymographs with high background signal will suffer from biased localization estimates. To disable bias correction, specify `bias_correction=False` to [`lk.track_greedy()`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.track_greedy.html#) and [`lk.refine_tracks_centroid()`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.refine_tracks_centroid.html).

> * The function [`ImageStack.define_tether()`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.ImageStack.html#lumicks.pylake.ImageStack.define_tether) now takes coordinates in image units (microns) rather than pixels.

Reviewer: ```suggestion
* The function [`ImageStack.define_tether()`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.ImageStack.html#lumicks.pylake.ImageStack.define_tether) now takes coordinates in calibrated spatial units (microns) rather than pixels. If calibration to microns is not available, units are still in pixels.
```

#### Other changes

* Removed all deprecated methods and properties from [`Kymo`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.kymo.Kymo.html), [`Scan`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.scan.Scan.html), [`PointScan`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.point_scan.PointScan.html), and [`ImageStack`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.ImageStack.html).
* Removed deprecated functions `filter_lines()`, `refine_lines_centroid()` and `refine_lines_gaussian()`.
* Removed deprecated property `lines` and method `save_lines` from `KymoWidgetGreedy`.
* Removed deprecated argument `exclude` from [`Kymo.line_timestamp_ranges()`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.kymo.Kymo.html#lumicks.pylake.kymo.Kymo.line_timestamp_ranges) and [`Scan.frame_timestamp_ranges()`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.scan.Scan.html#lumicks.pylake.scan.Scan.frame_timestamp_ranges). Use the argument `include_dead_time` instead.

> * Removed deprecated argument `line_width` from [`lk.track_greedy()`](https://lumicks-pylake.readthedocs.io/en/v1.0.0/_api/lumicks.pylake.track_greedy.html).

Reviewer: I put these as `Other changes` originally but I wonder if they should be listed under `Breaking changes`. Although I guess because we went through a deprecation pathway they're not technically breaking? thoughts?

Author: They are. I'll move them.
